### ISSUE TITLE
[Public API request] make parseForESLint accept ts.sourceFile 

### ISSUE THREAD

**Issue by Quramy** [typescript-eslint/typescript-eslint#774]
I want a public API for TypeScript language service plugin ( this is related to https://github.com/typescript-eslint/typescript-eslint/issues/254#issuecomment-515781974 ).

#### Signature

```ts
parseForESLintFromSourceFile(
    sourceFile: ts.sourceFile,
    options?: ParserOptions | null,
): ParseForESLintResult
```

#### Expected behavior
This new function should works almost the same as `parseForESLint`. However, the following points are different: 

- `result.ast` should be converted by `@typescript-eslint/typescript-estree/src/ast-converter` directly.
- `result.services.program` should be undefined ( I want to skip to call `createProgram`)

### Motivation

I develop a TypeScript language service plugin to integrate tsserver and ESLint, https://github.com/Quramy/typescript-eslint-language-service. 

In tsserver, ts.LanguageService does:

- tokenize source text and create an AST node. And the parsing is so fast because of ts.IncrementalParser.  
- provide ts.Program ( So ts-estree don't need call createProgram )

Of course, I can provide a language service plugin using only public `@typescript-eslint/parser` APIs. But they accept only string text. So source code's text would be tokenized twice if using these public APIs. It's so silly.

Now I use some internal APIs ( see https://github.com/Quramy/typescript-eslint-language-service/blob/v1.2.3/src/ast-converter.ts#L187 ) for the above performance reason. But depending on them is too fragile.

Please consider it.

Quramy

**Comment by bradzacher** [comment]
Happy to accept a PR for both cases!
It's not something we'll actively work on ourselves, as there's little call for it for the majority of the userbase, and we've got our hands full with everything else.

Quramy has built a ts language server plugin to coordinate linting via the language server instance, instead of installing an eslint plugin.

---

For the case of accepting a `ts.SourceCode` file, you'd essentially be wanting a thin wrapper on top of the conversion logic. I.e. you essentially want a function which does this, without the SourceCode creator:

https://github.com/typescript-eslint/typescript-eslint/blob/611dff3f58d3c91fd7d7ab9e07a536ae09f50522/packages/typescript-estree/src/parser.ts#L299-L309

---

For the case of sharing a `ts.Program`, this is a much larger, more difficult problem.
Our codebase is built around using `ts.createWatchProgram`, because it gives us a `BuilderProgram`, which is internally designed to be more amenable to the "incremental change" use case.

Our code (as of the commits I merged on the weekend) does a lot of manipulation of this program as well (in order to handle the IDE's persistent state usecase).

So you'd probably want something a bit different - a thin wrapper which accepts a set of programs ahead of the lint, and then when asked to parse a file, it iterates over the programs until it finds one that has the file, and then uses the same API above to convert the `SourceFile`.

**Comment by octogonz** [comment]
Cool. Should I open a separate issue for my scenario? I might work on this if I get some time.

**Comment by bradzacher** [comment]
You can if you want - there's no huge need to open a new issue.

**Comment by JoshuaKGoldberg** [comment]
I'll tackle the original request of making `parseForESLint` also accept a source file once #5834 is in.

**Comment by JoshuaKGoldberg** [comment]
#5892 was merged into the `v6` branch. This'll be available as part of our upcoming v6 major version! 🚀
